src/lib/openjp2/openjpeg.c

```diff
@@ -958,3 +1073,266 @@ opj_stream_t* OPJ_CALLCONV opj_stream_create_file_stream (
+		if (mode[1] == '+')
+			m = O_RDWR;
+		break;
+	case 'w':
+	case 'a':
+		m = O_RDWR | O_CREAT;
+		if (mode[0] == 'w')
+			m |= O_TRUNC;
+		break;
+	default:
+		break;
+	}
+	return m;
+}
+
+
+#ifdef _WIN32
+
+static OPJ_UINT64  opj_size_proc(opj_handle_t fd)
+{
+	ULARGE_INTEGER m;
+	m.LowPart = GetFileSize(fd, &m.HighPart);
+	return(m.QuadPart);
+}
+
+
+static void* opj_map(opj_handle_t fd, OPJ_SIZE_T len) {
+	void* ptr = NULL;
+	HANDLE hMapFile = NULL;
+
+	if (!fd || !len)
+		return NULL;
+
+	/* Passing in 0 for the maximum file size indicates that we
+	would like to create a file mapping object for the full file size */
+	hMapFile = CreateFileMapping(fd, NULL, PAGE_READONLY, 0, 0, NULL);
+	if (hMapFile == NULL) {
+		return NULL;
+	}
+	ptr = MapViewOfFile(hMapFile, FILE_MAP_READ, 0, 0, 0);
+	CloseHandle(hMapFile);
+	return ptr;
+}
+
+static OPJ_INT32 opj_unmap(void* ptr, OPJ_SIZE_T len){
+	OPJ_INT32 rc = -1;
+	if (ptr) 
+		rc = UnmapViewOfFile(ptr) ? 0 : -1; 
+	return rc;
+}
+
+static opj_handle_t opj_open_fd(const char* fname, const char* mode) {
+	void*	fd		= NULL;
+	OPJ_INT32 m		= -1;
+	DWORD			dwMode = 0;
+
+	if (!fname) 
+		return (void*)-1;
+	
+
+	m = opj_get_file_open_mode(mode);
+	switch (m) {
+	case O_RDONLY:
+		dwMode = OPEN_EXISTING;
+		break;
+	case O_RDWR:
+		dwMode = OPEN_ALWAYS;
+		break;
+	case O_RDWR | O_CREAT:
+		dwMode = OPEN_ALWAYS;
+		break;
+	case O_RDWR | O_TRUNC:
+		dwMode = CREATE_ALWAYS;
+		break;
+	case O_RDWR | O_CREAT | O_TRUNC:
+		dwMode = CREATE_ALWAYS;
+		break;
+	default:
+		return NULL;
+	}
+
+	fd = (opj_handle_t)CreateFileA(fname,
+		(m == O_RDONLY) ? GENERIC_READ : (GENERIC_READ | GENERIC_WRITE),
+		FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, dwMode,
+		(m == O_RDONLY) ? FILE_ATTRIBUTE_READONLY : FILE_ATTRIBUTE_NORMAL,
+		NULL);
+	if (fd == INVALID_HANDLE_VALUE) {
+		return (opj_handle_t)-1;
+	}
+	return fd;
+}
+
+static OPJ_INT32 opj_close_fd(opj_handle_t fd) {
+	OPJ_INT32 rc = -1;
+	if (fd) {
+		rc = CloseHandle(fd) ? 0 : -1;
+	}
+	return rc;
+}
+
+#else
+
+static OPJ_UINT64 opj_size_proc(opj_handle_t fd)
+{
+	struct stat sb;
+	if (!fd)
+		return 0;
+
+	if (fstat(fd, &sb)<0)
+		return(0);
+	else
+		return((OPJ_UINT64)sb.st_size);
+}
+
+static void* opj_map(opj_handle_t fd, OPJ_SIZE_T len) {
+	void* ptr = NULL;
+	OPJ_UINT64		size64 = 0;
+
+	if (!fd)
+		return NULL;
+
+	size64 = opj_size_proc(fd);
+	ptr = (void*)mmap(0, (size_t)size64, PROT_READ, MAP_SHARED, fd, 0);
+	return ptr == (void*)-1 ? NULL : ptr;
+}
+
+static OPJ_INT32 opj_unmap(void* ptr, OPJ_SIZE_T len) {
+	// unmap
+	if (ptr)
+		munmap(ptr, (off_t)len);
+	return 0;
+}
+
+static opj_handle_t opj_open_fd(const char* fname, const char* mode) {
+	opj_handle_t	fd = 0;
+	OPJ_INT32 m = -1;
+	if (!fname) {
+		return -1;
+	}
+	m = opj_get_file_open_mode(mode);
+	fd = open(fname, m, 0666);
+	if (fd < 0) {
+		if (errno > 0 && strerror(errno) != NULL) {
+			printf("%s: %s", fname, strerror(errno));
+		}
+		else {
+			printf("%s: Cannot open", fname);
```

Comment:
There should be no printf in the library code. Please use the existing message mechanism if necessary.